homeassistant/components/life360/helpers.py

```diff
@@ -1,7 +1,238 @@
 """Life360 integration helpers."""
-from life360 import Life360
 
+from __future__ import annotations
 
-def get_api(authorization=None):
+from collections.abc import Callable
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Any
+
+from life360 import Life360, Life360Error, LoginError
+
+from homeassistant.const import (
+    LENGTH_FEET,
+    LENGTH_KILOMETERS,
+    LENGTH_METERS,
+    LENGTH_MILES,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util.distance import convert
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    COMM_MAX_RETRIES,
+    COMM_TIMEOUT,
+    DOMAIN,
+    LOGGER,
+    SPEED_DIGITS,
+    SPEED_FACTOR_MPH,
+)
+
+
+@dataclass
+class IntegData:
+    """Integration data."""
+
+    cfg_options: dict[str, Any] | None = None
+    # ConfigEntry.unique_id: DataUpdateCoordinator
+    coordinators: dict[str, DataUpdateCoordinator] = field(
+        init=False, default_factory=dict
+    )
+    # member_id: ConfigEntry.unique_id
+    tracked_members: dict[str, str] = field(init=False, default_factory=dict)
+    logged_circles: list[str] = field(init=False, default_factory=list)
+    logged_places: list[str] = field(init=False, default_factory=list)
+
+    def __post_init__(self):
+        """Finish initialization of cfg_options."""
+        self.cfg_options = self.cfg_options or {}
+
+
+@dataclass
+class Life360Place:
+    """Life360 Place data."""
+
+    name: str
+    latitude: float
+    longitude: float
+    radius: float
+
+
+@dataclass
+class Life360Circle:
+    """Life360 Circle data."""
+
+    name: str
+    places: dict[str, Life360Place]
+
+
+@dataclass
+class Life360Member:
+    """Life360 Member data."""
+
+    # Don't include address field in eq comparison because it often changes (back and
+    # forth) between updates. If it was included there would be way more state changes
+    # and database updates than is useful.
+    address: str | None = field(compare=False)
+    at_loc_since: datetime
+    battery_charging: bool
+    battery_level: int
+    driving: bool
+    entity_picture: str
+    gps_accuracy: int
+    last_seen: datetime
+    latitude: float
+    longitude: float
+    name: str
+    place: str | None
+    speed: float
+    wifi_on: bool
+
+
+@dataclass
+class Life360Data:
+    """Life360 data."""
+
+    circles: dict[str, Life360Circle] = field(init=False, default_factory=dict)
+    members: dict[str, Life360Member] = field(init=False, default_factory=dict)
+
+
+def init_integ_data(
+    hass: HomeAssistant, cfg_options: dict[str, Any] | None = None
+) -> None:
+    """Initialize domain's hass data if necessary."""
+    hass.data.setdefault(DOMAIN, IntegData(cfg_options))
```

Comment:
Most of the data being stored is used across all config entries. That is necessary because, e.g., members can exist in multiple Life360 Circles accessed via different Life360 accounts (i.e., different config entries.) The only item that is config entry specific is the coordinator, which is why there is an item named coordinators whose key is the config entry ID. I could reorganize and add a "config_data" item under DOMAIN, i.e., a dictionary keyed by the config entry ID, but then it would have only one data item, namely coordinator. That seemed unnecessary. Basically the following:
```python
hass.data[DOMAIN].coordinators[entry.entry_id]
```
would have to change to:
```python
hass.data[DOMAIN].config_data[entry.entry_id].coordinator
```
That doesn't seem to be any better, in fact, I'd claim it was worse.

Reply:
Ok.